library/old_protocol/serialize.cpp

```diff
@@ -84,21 +84,53 @@ int serialize_lookup_request(dnet_net_state *st, std::unique_ptr<n2_request> msg
 
 int serialize_lookup_response(dnet_net_state *st, std::unique_ptr<n2_message> msg_in,
                               std::unique_ptr<n2_serialized> &out_serialized) {
-	auto impl = [&] {
-		auto &msg = static_cast<lookup_response &>(*msg_in);
+	auto &msg = static_cast<lookup_response &>(*msg_in);
 
-		msgpack::sbuffer msgpack_buffer;
-		msgpack::pack(msgpack_buffer, msg);
+	auto impl = [&] {
+		data_pointer data;
+		size_t size = 0;
+
+		if (msg.cmd.cmd == DNET_CMD_LOOKUP_NEW) {
+			msgpack::sbuffer msgpack_buffer;
+			msgpack::pack(msgpack_buffer, msg);
+
+			data = data_pointer::copy(msgpack_buffer.data(), msgpack_buffer.size());
+			size = msgpack_buffer.size();
+		} else if (msg.cmd.cmd == DNET_CMD_LOOKUP) {
+			auto path_size = msg.path.size() + 1;  // including 0-byte
+			size = sizeof(struct dnet_addr) + sizeof(struct dnet_file_info) + path_size;
+
+			auto addr = static_cast<dnet_addr *>(calloc(1, size));
+			if (addr == NULL) {
+				return -ENOMEM;
+			}
+			memcpy(addr, &st->n->addrs[0], sizeof(struct dnet_addr));
+			dnet_convert_addr(addr);
+
+			auto file_info = reinterpret_cast<dnet_file_info *>(addr + 1);
+			file_info->flen = path_size;
+			file_info->record_flags = msg.record_flags;
+			file_info->size = msg.data_size;
+			file_info->offset = msg.data_offset;
+			file_info->mtime = msg.data_timestamp;
+			if (msg.cmd.flags & DNET_FLAGS_CHECKSUM) {
+				memcpy(file_info->checksum, msg.data_checksum.data(), DNET_CSUM_SIZE);
+			}
+			memcpy(file_info + 1, msg.path.c_str(), path_size);
+			dnet_convert_file_info(file_info);
+
+			data = data_pointer::copy(addr, size);
```

Comment:
`data_pointer` could be allocated at the beginning and filled up in place:
```cpp
data = data_pointer::allocate(sizeof(struct dnet_addr) + sizeof(struct dnet_file_info) + path_size);

auto addr = data.data<dnet_addr>();
memcpy(addr, &st->n->addrs[0], sizeof(struct dnet_addr));
dnet_convert_addr(addr);
```

Reply:
5712430

Reply:
Yes, and I forgot to free memory. It's not needed with `data_pointer`.